### PR TITLE
Tests for QuickSort and DualQuickSort

### PR DESCRIPTION
**Pull request by AntonKluge**
**Link any issues if they have led to changes in the request**  

**Please describe the (sub-)exercises the tests are meant for**  
* GAD 6
* QuickSort tests
* DualQuickSort tests

**Please Note**
* It is a little overkill with JUnit, but it is more sorted when the tests run, it's easier to differentiate between tests, and it is pretty easy to add new test cases with individual parameters.
* Someone with a working implementation has to check DualQuickSortTest. The given Result checks if it sorts correctly and I tried to keep it as open as possible, but maybe there's another way to solve this (and pass Artemis tests) I don't know about.
* Sorry for the commit mess :)

### CODE REVIEW

File: GAD 6 - Verbessertes Sortieren/tests/simplesort/TesterDualQuicksortResult.java [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        }
+
+        Arrays.sort(startResult);
+        int[] copyEndResult = Arrays.copyOfRange(endResult, indicesOfLog[0], indicesOfLog[5] + 1);
+        Arrays.sort(copyEndResult);
+
+        assertEquals(startResult.length, copyEndResult.length, "Array: " + Arrays.toString(endResult) +
+                "\nIn range: " + indicesOfLog[0] + " to " + indicesOfLog[5] +
+            "\nThe arrays before and after the sort dont have the same length anymore, " +
+                    "you probably changed the to or from pointer");
```
Comment (CrsiX) [inline GAD 6 - Verbessertes Sortieren/tests/simplesort/TesterDualQuicksortResult.java:66]: Please fix the indent here too :)

File: GAD 6 - Verbessertes Sortieren/tests/simplesort/TesterDualQuicksortResult.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+            analyseLogsOfDualPivot(array);
+        }
+
+    }
+
+    private void analyseLogsOfDualPivot(int[] endResult) {
+        if (startResult == null || endResult == null) {
+            throw new IllegalArgumentException("Array: " + Arrays.toString(endResult) + "\n" +
+                                    "\nIn range: " + indicesOfLog[0] + " to " + indicesOfLog[5] + "\n" +
+                                "\nLogged partial Arrays without logging startResult or endResult");
```
Comment (CrsiX) [inline GAD 6 - Verbessertes Sortieren/tests/simplesort/TesterDualQuicksortResult.java:56]: Please fix the indent here :)

**Review by CrsiX — COMMENTED** [review]
Thanks for the work so far. :)

Some comments:
- I couldn't verify the correctness of the tests yet, but at least they seem to work, even though I think parts of the tests are designed too complicated. Tests should provide help to find bugs and should be as straightforward as possible imho. That's why I couldn't verify their correctness at first glance.
- All of the 28 tests of `DualPivotQuicksortTester` passed even though I had an **empty** `DualPivotQuicksort.sort` method. That's not what I would expect from a test suite. At least some test should fail when you don't sort anything, shouldn't they?
- Could you please align some of your indents "better", to improve readability? I have pointed out some places where they are somewhat visually misaligned.
- Usually, Java uses camelCase naming conventions, not snake_case (like e.g. Python). It would be nice if you could update your method names. :)

**Comment by AntonKluge** [comment]
> Thanks for the work so far. :)
> 
> Some comments:
> 
> * I couldn't verify the correctness of the tests yet, but at least they seem to work, even though I think parts of the tests are designed too complicated. Tests should provide help to find bugs and should be as straightforward as possible imho. That's why I couldn't verify their correctness at first glance.
> * All of the 28 tests of `DualPivotQuicksortTester` passed even though I had an **empty** `DualPivotQuicksort.sort` method. That's not what I would expect from a test suite. At least some test should fail when you don't sort anything, shouldn't they?
> * Could you please align some of your indents "better", to improve readability? I have pointed out some places where they are somewhat visually misaligned.
> * Usually, Java uses camelCase naming conventions, not snake_case (like e.g. Python). It would be nice if you could update your method names. :)

Thanks for the detailed feedback!!! 
I think I have solved most of it.

I totally understand the problem with complexity, my goal was to make the "tests sidebar" where you can see which tests pass and which don't, as clean and helpful as possible (and play a bit with JUnit, I'm new with tests). Also, it should be easy to add more tests following the same pattern. 

If you say this is still overkill and has no place in the repo, I totally understand! Maybe I'll find some time to make it simpler, or next week write the tests simpler from the beginning and try again.

**Comment by CrsiX** [comment]
Looks really good now, there are 20 failing tests out of 28 for my empty `sort` method (for reference: Artemis says I passed 4 out of 19 test cases, so that's probably okay).

Thanks for the additional clean-up. Even though I still haven't checked the 100% correctness of the tests: If someone else approves the correctness, I approve its merge. :)

> I totally understand the problem with complexity, my goal was to make the "tests sidebar" where you can see which tests pass and which don't, as clean and helpful as possible (and play a bit with JUnit, I'm new with tests). Also, it should be easy to add more tests following the same pattern.

That's great to hear. And I must admit that it looks well in my IDE. And easy extensibility is always a good thing.

> Maybe I'll find some time to make it simpler, or next week write the tests simpler from the beginning and try again.

Sure, go on :)

**Comment by babyygemperor** [comment]
Tests work perfectly! Amazing tests!
Thanks for contributing

**Pull request merged**